### PR TITLE
Update ConvertType

### PR DESCRIPTION
**Pull request by LeviPesin**
Related issue: https://github.com/mrdoob/three.js/pull/23601#issuecomment-1054057253

**Description**

* [ ] ~~Remove `valueComponents`.~~
* [ ] ~~Add arguments' flatting.~~
* [ ] ~~Remove `valueClass`.~~
* [x] Add cache.
* [x] Rework ConvertType.
* [x] Fix uint generation in NodeBuilder.

### CODE REVIEW

**Comment by LeviPesin** [comment]
Actually, here can be reused function `getValueFromType()` from https://github.com/mrdoob/three.js/pull/23663 to remove the `valueClass` parameter.

Also I suggest adding to the ConvertType some cache - so e.g. if `float(0)` would be called hundred times it will return the same value. This cannot be done with WeakMap, because it only accepts keys which are objects, so maybe it should be done with just Map.

**Comment by LeviPesin** [comment]
@sunag What is your opinion about something like `vec4(1.0, new Vector2(1.0, 1.0), 1.0)`? Should it be supported or rather not?

**Comment by sunag** [comment]
> @sunag What is your opinion about something like vec4(1.0, new Vector2(1.0, 1.0), 1.0)? Should it be supported or rather not?

Sounds good! 

Can you do this update without change `src`(core) code?

**Comment by LeviPesin** [comment]
This is the part that I called arguments' flatting. I changed the core code to remove valueComponents - I can revert it back.

**Comment by LeviPesin** [comment]
Done.

**Comment by sunag** [comment]
Can you check the `webgpu_nodes_playground` example? Seems that some vectors contains `NaN` values.

image

**Comment by LeviPesin** [comment]
I've fixed the error, but for some reason the models now appear to be white instead of black...

**Comment by sunag** [comment]
@LeviPesin Seems that `vec*()` no longer works with scalar values.

See `ColorSpaceNode.LinearTosRGB`:
```js
const a = sub( mul( pow( value.rgb, vec3( 0.41666 ) ), 1.055 ), vec3( 0.055 ) );
const b = mul( rgb, 12.92 );
const factor = vec3( lessThanEqual( rgb, vec3( 0.0031308 ) ) );

const rgbResult = mix( a, b, factor );

return join( rgbResult.r, rgbResult.g, rgbResult.b, value.a );
```

```glsl
nodeVar4 = pow( DiffuseColor.xyz, vec3<f32>( 0.0, 0.0, 0.0 ) );
nodeVar5 = ( nodeVar4 * vec3<f32>( 1.055 ) );
nodeVar6 = ( nodeVar5 - vec3<f32>( 0.0, 0.0, 0.0 ) );
nodeVar7 = ( DiffuseColor.xyz * vec3<f32>( 12.92 ) );
nodeVar8 = mix( nodeVar6, nodeVar7, vec3<f32>( 0.0, 0.0, 0.0 ) );
Output = vec4<f32>( nodeVar8.x, nodeVar8.y, nodeVar8.z, DiffuseColor.w );
```

**Comment by LeviPesin** [comment]
Fixed.

File: examples/jsm/nodes/ShaderNode.js [hunk 1]
```diff
@@ -208,13 +208,25 @@ export const label = ( node, name ) => {
 
 export const temp = ( node ) => nodeObject( new VarNode( nodeObject( node ) ) );
 
-const ConvertType = function ( type, valueClass = null, valueComponents = 1 ) {
+const ConvertType = function ( type, valueComponents = 1 ) {
+
+	const map = new ArrayMap();
```
Comment (sunag) [inline examples/jsm/nodes/ShaderNode.js:213]: Are you sure this approach will not generate `memory leaks`?

Comment (LeviPesin) [reply examples/jsm/nodes/ShaderNode.js:213]: Unfortunately it is impossible to use WeakMap here because each param can be primitive (and most often it will actually be). Maybe it is better to implement in ArrayMap some kind of deleting keys that are not used for a long time when its size exceeds some value?

Comment (LeviPesin) [reply examples/jsm/nodes/ShaderNode.js:213]: Something like this should work, I think...

Comment (sunag) [reply examples/jsm/nodes/ShaderNode.js:213]: This should also compose vectors of uniforms and variables too, just like we have in `JoinNode`.  e.g: `vec3( 1.0, colorNode )`
I would like to recommend a different approach to `ConvertType`:

1. Convert all parameters of `primitive values` and `Three.js values (Color, Vector2, ...)` to `ConstNode`.
2. Create a `JoinNode` and add all nodes.
3. Add a `ConvertNode` at the top to ensure the desired output.

**Comment by sunag** [comment]
This should also compose vectors of uniforms and variables too, just like we have in `JoinNode`.  e.g: `vec3( 1.0, colorNode )`
I would like to recommend a different approach to `ConvertType`:

1. Convert all parameters of `primitive values` and `Three.js values (Color, Vector2, ...)` to `ConstNode`.
2. Create a `JoinNode` and add all nodes.
3. Add a `ConvertNode` at the top to ensure the desired output.

File: examples/jsm/nodes/ShaderNode.js (hunk 1 above)
Comment (sunag) [reply examples/jsm/nodes/ShaderNode.js:213]: I think that we can try to follow an approach focused in `JoinNode` and remove `ArrayMap`. https://github.com/mrdoob/three.js/pull/23635#issuecomment-1073338514

Comment (sunag) [reply examples/jsm/nodes/ShaderNode.js:213]: About the cache, I think that we can create a `FloatNode` to most common values like: `-1`, `-.5`, `0`, `.5`, `1`. Other values do not tend to repeat themselves as often.

**Comment by LeviPesin** [comment]
I don't quite understand how it should work... Should `vec4( 1, new Vector3( 2, 3, 4 ) )` be `ConstNode( new Vector4( 1, 2, 3, 4 ) )` and `uniform( vec4( 1, new Vector3( 2, 3, 4 ) ) )` be `ArrayUniformNode( UniformNode( 1 ), UniformNode( new Vector3( 2, 3, 4 ) ) )`, or just `ArrayUniformNode( UniformNode( new Vector4( 1, 2, 3, 4 ) ) )`?

**Comment by sunag** [comment]
1. Convert all parameters of `primitive values` and `Three.js values (Color, Vector2, ...)` to `ConstNode`.
https://github.com/mrdoob/three.js/blob/74811ba4d28578caefbb6d5bcfc319194c16fabf/examples/jsm/nodes/ShaderNode.js#L213

Example 1
```js
// shader node api
const anyNode = normalView;
vec3( 10.0, anyNode );

// must be: node material api
const nodes = [ new ConstNode( 10.0 ), anyNode ];
```


Example 2
```js
// shader node declaration
vec3( 10.0, new Vector2( 0, 0 ) );

// must be: node material api
const nodes = [ new ConstNode( 10.0 ), new ConstNode( new Vector2( 0, 0 ) )  ];
```

2. Create a `JoinNode` and add all nodes.
3. Add a `ConvertNode` on top to ensure the desired output.
```js
const ConvertType = function ( type ) {
 ...
	return new ConvertNode( new JoinNode( nodes ), type );
}
```

**Comment by sunag** [comment]
Still need to consider the integers in the primitives [here](https://github.com/mrdoob/three.js/pull/23635#issuecomment-1073477866), but I think that's the way to go.

**Comment by LeviPesin** [comment]
OK, is it better now?

**Comment by sunag** [comment]
@LeviPesin Still about cache, I think that we can also replace `ArrayMap` to something more simple like [this](https://github.com/mrdoob/three.js/pull/23635#discussion_r830668205) ?
```js
// static cache of common values
const cache = [ .. float( 0 ), float( .5 ), float( 1 ) ];

// find a parameter that contains the same value
// let's consider only floats for now
const cacheNode = cache.find( ... ); 
```

**Comment by LeviPesin** [comment]
Is it better now?

Also, maybe it would be better to remove `PI`, `PI2`, `PI_HALF`, `RECIPROCAL_PI`, `RECIPROCAL_PI2` (but keep `EPSILON` and add `INFINITY = 1 / EPSILON`)? Just something like `Math.PI` or `float( Math.PI )` could be used instead when needed - and it will cache.

Also, the cache should be added to the `new ConstNode( Number( prop ), 'uint' )` and `new ShaderNodeObject( new ConstNode( obj ) );`, I think.

File: examples/jsm/nodes/ShaderNode.js [hunk 2]
```diff
@@ -186,7 +186,7 @@ export const nodeObject = ( val ) => {
 
 export const uniform = ( constNode ) => {
 
-	return nodeObject( new UniformNode( constNode.value, constNode.nodeType ) );
+	return nodeObject( new UniformNode( constNode.value?.clone?.() || constNode.value, constNode.nodeType ) );
```
Comment (sunag) [inline examples/jsm/nodes/ShaderNode.js]: Why did you need to clone this class?

**Comment by sunag** [comment]
> Also, maybe it would be better to remove PI, PI2, PI_HALF, RECIPROCAL_PI, RECIPROCAL_PI2 (but keep EPSILON and add INFINITY = 1 / EPSILON)? Just something like Math.PI or float( Math.PI ) could be used instead when needed - and it will cache.

@LeviPesin I think we can keep as is for now...

File: examples/jsm/nodes/ShaderNode.js (hunk 2 above)
Comment (LeviPesin) [reply examples/jsm/nodes/ShaderNode.js]: I think it can be removed, I have added it because of cache.

File: examples/jsm/nodes/ShaderNode.js [hunk 3]
```diff
@@ -208,66 +211,83 @@ export const label = ( node, name ) => {
...
+		if ( params.length === 0 ) {
 
-		if ( ( params.length === 1 ) && ( valueComponents !== 1 ) ) {
+			node = nodeObject( new ConstNode( getValueFromType( type ), type ) );
 
-			// Providing one scalar value: This value is used for all components
+		} else {
 
-			for ( let i = 1; i < valueComponents; i ++ ) {
+			if ( type === 'color' && params.length === 1 ) {
```
Comment (LeviPesin) [inline examples/jsm/nodes/ShaderNode.js:228]: See https://github.com/mrdoob/three.js/issues/23634.

File: examples/jsm/nodes/ShaderNode.js [hunk 4]
```diff
@@ -186,7 +186,10 @@ export const nodeObject = ( val ) => {
 
 export const uniform = ( constNode ) => {
 
-	return nodeObject( new UniformNode( constNode.value, constNode.nodeType ) );
+	// TODO: get constNode from .traverse() in the future
+	const value = constNode.value === undefined ? constNode.node?.value : constNode.value;
```
Comment (LeviPesin) [inline examples/jsm/nodes/ShaderNode.js:190]: Why is this needed?

Comment (sunag) [reply examples/jsm/nodes/ShaderNode.js:190]: In case of `ConstNode` under a `ConvertNode`.

File: examples/jsm/nodes/ShaderNode.js [hunk 5]
```diff
@@ -231,26 +233,46 @@ const ConvertType = function ( type ) {
...
 
 			} );
 
 			node = nodeObject( new ConvertNode( nodes.length === 1 ? nodes[ 0 ] : new JoinNode( nodes ), type ) );
 
 		}
 
+		if ( params.length === 1 && typeof params[ 0 ] !== 'object' ) {
+
+			nodesCacheMap.set( params[ 0 ], node );
```
Comment (sunag) [inline examples/jsm/nodes/ShaderNode.js:272]: @LeviPesin The idea is to avoid dynamic cache storage because it's a global cache.

Comment (LeviPesin) [reply examples/jsm/nodes/ShaderNode.js:272]: Why there cannot be a static cache for parameters passed and a dynamic cache per type for caching values like `float(1)`, `float(0)`, and `int(0)`?

Comment (sunag) [reply examples/jsm/nodes/ShaderNode.js:274]: We can have a `cacheLib = { float : new Map(), 'int' .. }` if necessary, but I would avoid dynamic strong storage here.

Comment (LeviPesin) [reply examples/jsm/nodes/ShaderNode.js:276]: > We can have a cacheLib = { float : new Map(), 'int' .. }

What would be the difference from the dynamic storage?

File: examples/jsm/nodes/ShaderNode.js (hunk 3 above)
Comment (LeviPesin) [reply examples/jsm/nodes/ShaderNode.js:256]: So should it actually be `params.length === 3` here?

File: examples/jsm/nodes/ShaderNode.js (hunk 5 above)
Comment (sunag) [reply examples/jsm/nodes/ShaderNode.js:276]: Store a thousand different values and you never can reduce the cache again because it's a global strong reference, so let's I think that we keep it storage static just to common values.

Comment (LeviPesin) [reply examples/jsm/nodes/ShaderNode.js:276]: > Store a thousand different values and you never can reduce the cache again because it's a global strong reference

But that would be also true for any Map (which is not WeakMap), wouldn't it be? In both cases we have `numberOfTypes` maps which keys are primitive (and so their keys and values would not be GC'd in any case, I think...).

Comment (sunag) [reply examples/jsm/nodes/ShaderNode.js:276]: for this reason we should fixed only these values https://github.com/mrdoob/three.js/pull/23635#discussion_r830668205

File: examples/jsm/nodes/ShaderNode.js (hunk 3 above)
Comment (sunag) [reply examples/jsm/nodes/ShaderNode.js:256]: the use of `color()` should be `color( 0xff00ff )`

Comment (LeviPesin) [reply examples/jsm/nodes/ShaderNode.js]: Then I think it should be [this](https://github.com/mrdoob/three.js/pull/23635/commits/1c0494f5b5af1bac5e876136fa5526b05a9f965a).

File: examples/jsm/nodes/ShaderNode.js (hunk 5 above)
Comment (LeviPesin) [reply examples/jsm/nodes/ShaderNode.js:278]: So, you are suggesting to limit the `constNodesCacheMap` to only common float values and `true` and false, and to limit the `nodesCacheMap` depending on type - common floats for float, common ints for int, common uints for uint, and common bools for bool (and nothing for other types)?

Comment (LeviPesin) [reply examples/jsm/nodes/ShaderNode.js:278]: Is it better now?

Comment (sunag) [reply examples/jsm/nodes/ShaderNode.js]: Yeess. Looks great! 👍

**Comment by sunag** [comment]
Thanks! 👍
